packages/playwright-core/src/server/debugController.ts

```diff
@@ -62,7 +63,10 @@ export class DebugController extends SdkObject {
   setReportStateChanged(enabled: boolean) {
     if (enabled && !this._trackHierarchyListener) {
       this._trackHierarchyListener = {
-        onPageOpen: () => this._emitSnapshot(false),
+        onPageOpen: page => {
+          this._emitSnapshot(false);
+          page.on(Page.Events.InternalFrameNavigatedToNewDocument, () => this._emitSnapshot(false));
```

Comment:
- We should only `emitSnapshot` upon main frame navigation to avoid spamming.
- We have to remove the listener from the page/pages at some point.